### ISSUE TITLE
External Datasource returning a list of DataNode and Jax data_iterator

### ISSUE THREAD

**Issue by clementpoiret** [NVIDIA/DALI#5614]
### Describe the question.

HI, I am trying to use a dataset containing AVIF images. Therefore, I followed the tutorials to setup an External Datasource, returning both images and labels. However, as I use Jax, if I want to setup a data_iterator, I have to set a reader name so that the decorator passes sharding details to the reader, such as:

```python
import nvidia.dali.fn as fn
from nvidia.dali.plugin.jax import data_iterator


@data_iterator(output_map=["images", "labels"], reader_name="image_reader")
def iterator_fn():
    jpegs, labels = fn.readers.file(file_root=image_dir, name="image_reader")
    images = fn.decoders.image(jpegs, device="mixed")
    images = fn.resize(images, resize_x=256, resize_y=256)
    return images, labels


iterator = iterator_fn(batch_size=8)
```

However, for an `external_source`, setting `name` makes the reader return a single DataNode. Here is my minimal reproductible example:

```python
class ExternalInputIterator(object):
    def __init__(
        self, batch_size: int, images_dir: str, pattern: str = "**/images/*.avif"
    ):
        self.images_dir = Path(images_dir)
        self.batch_size = batch_size

        self.images = list(self.images_dir.glob(pattern))

        shuffle(self.images)

    def __iter__(self):
        self.i = 0
        self.n = len(self.images)
        return self

    def __next__(self):
        batch = []
        labels = []  # Placeholder if we need it
        for _ in range(self.batch_size):
            avif_filename = self.images[self.i]

            image = np.array(Image.open(avif_filename), dtype=np.uint8)
            batch.append(image)
            labels.append(np.array([0], dtype=np.uint8))

            self.i = (self.i + 1) % self.n

        return (batch, labels)


batch_size = 16
eii = ExternalInputIterator(batch_size, "/mnt/hdd/datasets/avif_ds")


@data_iterator(output_map=["images", "labels"], reader_name="image_reader")
def rf1_iterator():
    avifs, labels = fn.external_source(
        source=eii, name="image_reader", dtype=types.UINT8
    )
    images = fn.decoders.image(avifs, device="mixed", output_type=types.RGB)
    images = fn.resize(images, resize_x=224, resize_y=224)
    return images, labels


iterator = rf1_iterator(batch_size=batch_size)
```

which returns the following error:

```
def iterator_fn():  File "dataloader.py", line 53, in rf1_iterator
    avifs, labels = fn.external_source(
    ^^^^^^^^^^^^^
  File ".venv/lib/python3.12/site-packages/nvidia/dali/_utils/hacks.py", line 23, in __iter__
    raise TypeError("The objects of type `", type(self), "` are not iterable.")
TypeError: ('The objects of type `', <class 'nvidia.dali.pipeline.DataNode'>, '` are not iterable.')
```

Do you have any workaround to make the pipeline work?

Thanks!

### Check for duplicates

- [x] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by JanuszL** [comment]
Hi @clementpoiret,

Thank you for reaching out. The `external_source` doesn't work with `name` as the number of iterations is not predefined and the `external_source` can raise a StopIteration exception anytime, and it can be conditioned on things beyond the pipeline knowledge/control. If you know the side of your data set feel free to set the `size` argument of the data iterator to a fixed value.
Please let us know if it works for you.

**Comment by awolant** [comment]
Hello @clementpoiret 

thanks for asking the question. The particular error you are seeing is caused by how your `ExternalInputIterator` is created and not interaction with JAX `data_iterator`.

In addition, as @JanuszL mentioned, you can control the size of the iterator with `size` argument. Otherwise, you can get endless iterator and control it from outside (like `range` or any other way). 

I prepared an example to show how to get functional JAX iterator with external source: 

```python
import nvidia.dali.fn as fn
from nvidia.dali.plugin.jax import data_iterator

import numpy as np

import types
import numpy as np
from random import shuffle
import nvidia.dali.fn as fn
import nvidia.dali.types as types


class ExternalInputIterator(object):
    def __init__(self, batch_size):
        self.images_dir = "data/images/"
        self.batch_size = batch_size
        with open(self.images_dir + "file_list.txt", "r") as f:
            self.files = [line.rstrip() for line in f if line != ""]
        shuffle(self.files)

    def __iter__(self):
        self.i = 0
        self.n = len(self.files)
        return self

    def __next__(self):
        batch = []
        labels = []
        for _ in range(self.batch_size):
            jpeg_filename, label = self.files[self.i].split(" ")
            f = open(self.images_dir + jpeg_filename, "rb")
            batch.append(np.frombuffer(f.read(), dtype=np.uint8))
            labels.append(np.array([label], dtype=np.uint8))
            self.i = (self.i + 1) % self.n
        return (batch, labels)


batch_size = 16
eii = ExternalInputIterator(batch_size)


@data_iterator(output_map=["images", "labels"], size=100)
def pipe_fn():
    jpegs, labels = fn.external_source(source=eii, num_outputs=2, dtype=types.UINT8)
    images = fn.decoders.image(jpegs, device="mixed", output_type=types.RGB)
    images = fn.resize(images, resize_x=224, resize_y=224)
    return (images, labels)


iterator = pipe_fn(batch_size=batch_size, num_threads=2, device_id=0)

for batch in iterator:
    print(batch["images"].shape)
```

You can run it with our test data from [here](https://github.com/NVIDIA/DALI/tree/main/docs/examples/data). Let me know if that helps and if you have any other questions.

**Comment by clementpoiret** [comment]
Nice, thanks for your help @JanuszL and @awolant! That's exactly what I was trying to do!

**Comment by awolant** [comment]
Your welcome. Could I ask what motivated you to try DALI with JAX? Did you experience any data bottlenecks in JAX workflow? Or were you already DALI user and you are migrating to JAX from other framework? 

I am asking because we would love to get more feedback about DALI and JAX integration to guide future development of this feature. Thanks a lot for your help.

**Comment by clementpoiret** [comment]
Most of the time I use PyTorch.
For a work project I decided to give Jax a try, and because Jax doesn't have built-in data management, it feels weird to have to use either PT's DataLoaders or tensorflow's `tf.data`.
So I searched a more "focused" tool for data loading and transformations, and found out DALI, which looks promising 👍 
Jax is clearly awesome, but because it is not as popular as PyTorch and TF, it's a bit hard to find well-maintained third-party libraries, so I feel Jax directly benefits from DALI!
